### PR TITLE
Fix cupy.take to become same as nupy.take(mode='raise')

### PR DESCRIPTION
**Pull request by kikusu**
This PR fixed some problem.
- different axis out of range error
```
In [17]: cupy.arange(24).reshape(2,3,4).take(1, axis=4)
ValueError: Axis overrun

In [18]: cupy.arange(24).reshape(2,3,4).take(1, axis=-4)
IndexError: tuple index out of range
```

- indicies < 0
```
In[24] cupy.arange(24).reshape(2,3,4).take([-1], axis=0)
Out[24]:
array([[[-9221023477734834572,   433521081881591445,   576601490364530678,
          4756136008902370279],
        [-2017608234743366152, -8070378514218255015,       71469669023658,
          2522101553271406586],
        [ 3530895775509880794,  3170538536574384045,  2594660525081321470,
          1315125858657353719]]])
```

- different out param shape
```
In [25]: a = cupy.array(0)

In [26]: cupy.arange(24).reshape(2,3,4).take(1, out=a)
Segmentation fault (core dumped)

In [2]: a = cupy.array([0])

In [3]: cupy.arange(24).reshape(2,3,4).take(1, out=a)
Out[3]: array([1])

In [4]: cupy.arange(24).reshape(2,3,4).take(1)
Out[4]: array(1)
```

this PR add
- minus axis range check
- minus indices range check
- out type, shape check when indices is scala

The range check makes it the same nupy.take(mode='raise')

### CODE REVIEW

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -1698,11 +1706,21 @@ cpdef ndarray _take(ndarray a, indices, axis=None, ndarray out=None):
...
+            if out.dtype != a.dtype:
+                raise TypeError('Output dtype mismatch')
+            if out.shape != a.shape[1:]:
+                raise ValueError('Output shape mismatch')
+            out[()] = a[indices]
             return out
     elif not isinstance(indices, ndarray):
         indices = array(indices, dtype=int)
 
+    if not ((-index_range <= indices) & (indices < index_range)).all():
```
Comment (okuta) [inline cupy/core/core.pyx]: This check is to decrease the performance.
Please comment out this code.
We may provide a debug mode.

File: cupy/core/core.pyx [hunk 2]
```diff
@@ -1698,11 +1706,21 @@ cpdef ndarray _take(ndarray a, indices, axis=None, ndarray out=None):
...
+        msg = "Index %s is out of bounds for axis %s with size %s" \
+              % (indices, axis, index_range)
+        raise IndexError(msg)
+    indices %= index_range
```
Comment (okuta) [inline cupy/core/core.pyx]: How about combined with `_take_kernel`?

File: tests/cupy_tests/core_tests/test_ndarray.py [hunk 3]
```diff
@@ -86,25 +86,85 @@ def test_take(self, xp, dtype):
...
+    shape = ()
+
+    @testing.for_all_dtypes()
+    @testing.numpy_cupy_array_equal(accept_error=False)
+    def test_take(self, xp, dtype):
+        a = testing.shaped_arange(self.shape, xp, dtype)
+        r1 = a.take(self.indices, self.axis)
+        r2 = xp.zeros_like(r1)
+        a.take(self.indices, self.axis, out=r2)
+        assert (r1 == r2).all()
```
Comment (okuta) [inline tests/cupy_tests/core_tests/test_ndarray.py]: Please use `self.assertTrue` or `numpy.testing.assert_array_equal`.

File: tests/cupy_tests/core_tests/test_ndarray.py [hunk 4]
```diff
@@ -86,25 +86,85 @@ def test_take(self, xp, dtype):
...
+    shape = (3, 4, 5)
+
+    @testing.for_all_dtypes()
+    @testing.numpy_cupy_array_equal(accept_error=False)
+    def test_take(self, xp, dtype):
+        a = testing.shaped_arange(self.shape, xp, dtype)
+        r1 = a.take(self.indices, self.axis)
+        r2 = xp.zeros_like(r1)
+        a.take(self.indices, self.axis, out=r2)
+        assert (r1 == r2).all()
```
Comment (okuta) [inline tests/cupy_tests/core_tests/test_ndarray.py]: Please use `self.assertTrue` or `numpy.testing.assert_array_equal`.

**Comment by kikusu** [comment]
Thank you for review.
- I removed indices check.
  - Default mode is numpy.take(mode="wrap")
- Test code fixed used cupy.testing.assert_array_equal.
- Indices check conbined with _take_kernel

File: cupy/core/core.pyx [hunk 5]
```diff
@@ -1660,45 +1660,61 @@ right_shift = _create_bit_op(
 # -----------------------------------------------------------------------------
 
 cdef _take_kernel = ElementwiseKernel(
-    'raw T a, S indices, int32 cdim, int32 rdim, int32 adim',
+    'raw T a, S indices, int32 cdim, int32 rdim, int32 adim, S index_range',
     'T out',
     '''
+      S wrap_indices = (indices - _floor_divide(indices, index_range) * index_range) * (index_range != 0);
```
Comment (okuta) [inline cupy/core/core.pyx]: How about following code?
```C
S wrap_indices = indices % index_range;
if (wrap_indices < 0) wrap_indices += index_range;
```

Comment (kikusu) [inline cupy/core/core.pyx]: Sorry, I was convinced that `indices` was array. I fixed it.

**Comment by okuta** [comment]
Thank you!

**Pull request merged**